lang/src/main/kotlin/org/partiql/lang/util/NumberExtensions.kt

```diff
@@ -256,3 +257,302 @@ val Number.isPosInf get() = when (this) {
+ * See https://dl.acm.org/doi/pdf/10.1145/214408.214413
+ */
+fun BigDecimal.squareRoot(mathContext: MathContext = MATH_CONTEXT): BigDecimal {
+    if (this.signum() < 0) {
+        throw ArithmeticException("Cannot take root of a negative number")
+    }
+
+    // We want to utilize the floating number's sqrt method to take an educated guess
+    // to make sure the number is representable
+    // we operate on normalized mantissa, which is [0.1, 10)
+    val stripped = this.stripTrailingZeros()
+    val scale = stripped.scale() - stripped.precision() + 1
+    val scaleAdj = if (scale % 2 == 0) scale else scale - 1
+    val mantissa = stripped.scaleByPowerOfTen(scaleAdj)
+
+    val guess = BigDecimal.valueOf(kotlin.math.sqrt(mantissa.toDouble()))
+
+    // Conservative guess of the precision of the result of a floating point calculation.
+    var guessPrecision = 10
+
+    // we need this additional logic in case of overflow
+    val targetPrecision = mathContext.precision
+    val normalizedPrecision = mantissa.precision()
+    var approx = guess
+
+    val zeroPointFive = BigDecimal.ONE.divide(BigDecimal.valueOf(2))
+    do {
+        // plus 2 for precision buffering
+        val operatingPrecision = kotlin.math.max(
+            kotlin.math.max(guessPrecision, targetPrecision + 2),
+            normalizedPrecision
+        )
+        val tempMC = MathContext(operatingPrecision, RoundingMode.HALF_EVEN)
+        approx = zeroPointFive.multiply(approx.add(mantissa.divide(approx, tempMC), tempMC))
+        // the magic number here is 2p + 2, consider precision(x*x) is maxed at precision(x) + precision(x)
+        guessPrecision = 2 * guessPrecision + 2
+    } while (guessPrecision < targetPrecision + 2)
+
+    // scale modification
+    val unModifiedRes = approx.scaleByPowerOfTen(-scaleAdj / 2).round(MATH_CONTEXT)
+
+    return unModifiedRes.roundToDigits(mathContext)
+}
+
+/**
+ * Computes e^x of a given BigDecimal x.
+ */
+fun BigDecimal.exp(mathContext: MathContext = MATH_CONTEXT): BigDecimal {
+    val operationMC = MathContext(
+        if (2 * mathContext.precision < 0) Int.MAX_VALUE else 2 * mathContext.precision,
+        mathContext.roundingMode
+    )
+    return if (this.signum() == 0) {
+        BigDecimal.ONE.roundToDigits(mathContext)
+    } else if (this.signum() == -1) {
+        val reciprocal = this.negate().expHelper(operationMC)
+        BigDecimal.valueOf(1)
+            .divide(
+                reciprocal,
+                operationMC
+            ).roundToDigits(mathContext)
+    } else {
+        this.expHelper(operationMC).roundToDigits(mathContext)
+    }
+}
+
+/**
+ * Computes the exponential value of a BigDecimal.
+ */
+private fun BigDecimal.expHelper(mathContext: MathContext): BigDecimal {
+    // For faster convergence, we break exponent into integer and fraction parts.
+    // e^x = e^(i+f) = (e^(1+f/i)) ^i
+    // 1 + f/i < 2
+    var intPart = this.divideToIntegralValue(BigDecimal.ONE).setScale(0)
```

Comment:
Would this be simpler?
```suggestion
    var intPart = this.toBigInteger()
```

Reply:
Looking into this, the initial thought was, I was not confident on converting big decimal to big Integer then to big decimal will be faster.